src/jit/morph.cpp

```diff
@@ -1242,158 +1242,96 @@ void fgArgInfo::RemorphReset()
     {
         GenTree* argx     = nullptr;
         unsigned regIndex = 0;
-
-        /* process the register argument list */
-        for (GenTreeArgList* list = callTree->gtCall.gtCallLateArgs; list; (regIndex++, list = list->Rest()))
+        if (reMorphing)
         {
-            argx = list->Current();
-            assert(!argx->IsArgPlaceHolderNode()); // No place holders nodes are in gtCallLateArgs;
-            if (regIndex == regArgInx)
+            // Find the arg in the late args list.
+            unsigned regArgInx = curArgTabEntry->lateArgInx;
+            for (GenTreeArgList* list = callTree->gtCall.gtCallLateArgs; list; (regIndex++, list = list->Rest()))
             {
-                break;
+                argx = list->Current();
+                assert(!argx->IsArgPlaceHolderNode()); // No place holders nodes are in gtCallLateArgs;
+                if (regIndex == regArgInx)
+                {
+                    break;
+                }
+            }
+            assert(regIndex == regArgInx);
+            if (curArgTabEntry->node != argx)
+            {
+                curArgTabEntry->node = argx;
             }
         }
-        assert(regIndex == regArgInx);
-        assert(regArgInx == curArgTabEntry->lateArgInx);
-
-        if (curArgTabEntry->node != argx)
+        else
         {
-            curArgTabEntry->node = argx;
+            assert(curArgTabEntry->lateArgInx == (unsigned)-1);
+            curArgTabEntry->node = node;
         }
     }
     return curArgTabEntry;
 }
 
-void fgArgInfo::RemorphStkArg(unsigned argNum, GenTree* node, GenTree* parent, unsigned numSlots, unsigned alignment)
+void fgArgInfo::UpdateStkArg(fgArgTabEntry* curArgTabEntry, GenTree* node, bool reMorphing)
 {
-    fgArgTabEntry* curArgTabEntry = nullptr;
-    bool           isRegArg       = false;
-    unsigned       regArgInx      = 0;
-    GenTree*       argx;
-    unsigned       inx;
-
-    for (inx = 0; inx < argCount; inx++)
-    {
-        curArgTabEntry = argTable[inx];
-
-        if (curArgTabEntry->parent != nullptr)
-        {
-            assert(curArgTabEntry->parent->OperIsList());
-            argx     = curArgTabEntry->parent->Current();
-            isRegArg = (argx->gtFlags & GTF_LATE_ARG) != 0;
-        }
-        else
-        {
-            argx     = curArgTabEntry->node;
-            isRegArg = true;
-        }
-
-        if (curArgTabEntry->argNum == argNum)
-        {
-            break;
-        }
-
-        if (isRegArg)
-        {
-            regArgInx++;
-        }
-    }
+    bool isLateArg = (curArgTabEntry->lateArgInx != (unsigned)-1);
 
-    nextSlotNum = roundUp(nextSlotNum, alignment);
-
-    assert(curArgTabEntry->argNum == argNum);
+    noway_assert(curArgTabEntry->parent != nullptr);
+    assert(curArgTabEntry->parent->OperIsList());
+    assert(curArgTabEntry->parent->Current() == node);
+    nextSlotNum = (unsigned)roundUp(nextSlotNum, curArgTabEntry->alignment);
```

Comment:
Would it be possible (and useful) to remove (or hoist to caller?) the `nextSlotNum` side-effect? That seems potentially confusing/dangerous.

Reply:
This is pre-existing. It could probably be changed, but would require some refactoring, as `nextSlotNum` is a (private) member of `fgArgInfo`. My thought is that we should leave it as-is.